### PR TITLE
Implement label control pairs

### PR DESCRIPTION
**Pull request by tomalec**
https://github.com/Starcounter/uniform.css/issues/29

### CODE REVIEW

**Comment by alshakero** [comment]
The margin is 7px more than it should be
image

File: index.html [hunk 1]
```diff
@@ -309,10 +309,11 @@ <h3 class="app-header"><slot name="app/content"></slot></h3>
             </div>
         </section>
         <section>
-            <h2>Form control</h2>
+            <h2>Label control</h2>
             <code>.uni-label-control</code>
-            <p>Puts slotted input in the left column, everything else in the right column.</p>
+            <p>Puts slotted label to the top row, everything else in the bottom row.</p>
             <p>It works for distributed elements.</p>
+            <p>Could tak optional `.uni-label-control-close` to distribute close button to the top right area.</p>
```
Comment (alshakero) [inline index.html]: Could take?

**Comment by tomalec** [comment]
> The margin is 7px more than it should be

From the designs it's hard to guess where to count from. From the bottom of `l` or should it be from the bottom of `j`, or just from the bottom of text line.

I'm picking the last one, then given by underwear `0.5rem (5px)` of margin-bottom of `label`, I'll change grid-gap to another `0.5rem (5px)`, so we should end up with 10px from end of text line to input

**Comment by tomalec** [comment]
BTW, Should underwer.css even add this `margin-bottom` for `label` and `label>span`?
https://github.com/Starcounter/underwear.css/blob/master/src/underwear.css#L225-L235

In my opinion, it should even care about `label>span`. Then speaking of the label, underwear should either do nothing, then uniform would add it from the shadow, or it should add the margin, but then we should be fie that for shadow roots without underwear inside, there will be no such margin. So we need to decide whether the numbers given by @andwah applies for 
- `Underwear in light, uniform in shadow & all elements in light`,
- `Underwear in light, uniform in shadow, underwear in shadow & elements wherever`,

Supporting the case `Underwear in light, uniform in shadow & elements wherever` would require us to include entire underwear in uniform.

//cc @warpech

File: index.html [hunk 2]
```diff
@@ -348,13 +351,41 @@ <h2>Form control</h2>
...
+            <pre><code>&lt;shadowroot mode="open"&gt;
+    &lt;style&gt;@import 'uniform.css';&lt;/style&gt;
+    &lt;div class="uni-label-control"&gt;
+        &lt;slot name="myapp/label">&lt;/slot&gt;
+        &lt;div class="uni-label-control-close"&gt;&lt;slot name="myapp/close"&gt;&lt;/slot&gt;&lt;/div&gt;
+        &lt;slot&gt;&lt;/slot&gt;
+    &lt;/div&gt;
+&lt;/shadowroot&gt;
+&lt;label slot="myapp/label"&gt;Label&lt;/label&gt;
+&lt;span slot="myapp/close" title="close"&gt;🗙&lt;/span&gt;
```
Comment (PatrykSzwer) [inline index.html]: Please remove this x button from the demo. The only function of it It's making the whole case confusing.

File: uniform.css [hunk 3]
```diff
@@ -87,27 +87,44 @@
...
+                 "input input"
+                 /  1fr auto;
+  grid-gap: 0.5rem;
   align-items: baseline;
 }
 
 .uni-label-control > ::slotted(*) {
-  grid-column: 2 / 3;
+  grid-area: input;
   width: 100%;
```
Comment (PatrykSzwer) [inline uniform.css]: Is this width: 100% necessary? With this when I'm trying to use uni-label-control from this:

**Comment by tomalec** [comment]
Do you want to remove the button just from the demos, or from grid layout at all?

**Comment by tomalec** [comment]
Also, I'd really appreciate if we could decide on uniform vs underwear+uniform approach, as stated in https://github.com/Starcounter/uniform.css/pull/42#issuecomment-366445107

But that's a general problem not just for alerts.

**Comment by alshakero** [comment]
> Also, I'd really appreciate if we could decide on uniform vs underwear+uniform approach, as stated in #42 (comment)

I totally agree with that comment. I'm not comfortable with `underwear` messing with layout. I'll add a commit to my latest PR.

**Comment by alshakero** [comment]
Done in https://github.com/Starcounter/underwear.css/pull/47/commits/e4428ae0d372281c3e9049fbe74d4445da00f87c

**Comment by warpech** [comment]
> Supporting the case `Underwear in light, uniform in shadow & elements wherever` would require us to include entire underwear in uniform.

If the app developer wants to use `label` elements in Shadow DOM, he can import `underwear.css` in Shadow DOM -- just as he would do it to use `button` elements in Shadow DOM.

----

@tomalec if there is still some conceptual discussion to have about margins, please make an issue about this. We can spend any other day discussing margins. Overall, I think that Underwear.css should not change the default `display` of elements, so I support @alshakero's change linked above.

File: uniform.css [hunk 4]
```diff
@@ -86,28 +86,39 @@
  */
 
 .uni-label-control {
-  display: grid;
-  grid-template-columns: 150px 1fr;
-  grid-column-gap: 10px;
+  display: inline-grid;
```
Comment (PatrykSzwer) [inline uniform.css]: I think previous `display: grid;` worked better. Other than that changes looks good.

Comment (tomalec) [reply uniform.css]: Why do you think it worked better, that's the only way not to have not-full width experience that you requested.

**Comment by tomalec** [comment]
@warpech  I created a separate issue about the conceptual approach to the design from underwear and uniform spread across shadow dom boundary https://github.com/Starcounter/uniform.css/issues/45

What we have left here is the discussion on `display` property of `label-control` and probably update uniform for @alshakero changes.

File: uniform.css (hunk 4 above)
Comment (PatrykSzwer) [reply uniform.css]: ? I've tried to use it with display: grid; and nothing happen to width:

Comment (tomalec) [reply uniform.css]: Now, all your wrappers take 100%. If that's desired, I'm changing it back to the `grid`.

**Comment by andwah** [comment]
In Current [view](https://starcounter.github.io/uniform.css/) the label position is off by (around 6px) on y-axis
image
Red color is the wanted design

**Comment by tomalec** [comment]
Please specify where you count 10px from? baseline, or line-box (https://iamvdo.me/en/blog/css-font-metrics-line-height-and-vertical-align)

I specified it to counted from (font-relative) line-box to the input rectangle. And used `rem` unit to be relative to the font size, therefore more responsive.

**Comment by warpech** [comment]
I've created a new issue for it: https://github.com/Starcounter/uniform.css/issues/54

**Pull request merged**